sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuWindowInPandasExecBase.scala

```diff
@@ -78,96 +77,265 @@ abstract class GpuWindowInPandasExecMetaBase(
+    partitionRefs: Seq[Expression],
+    isSplitGroups: Boolean,
+    windowFrames: Seq[GpuSpecifiedWindowFrame],
     inputRows: GpuMetric,
     inputBatches: GpuMetric,
-    spillCallback: RapidsBuffer.SpillCallback) extends Iterator[ColumnarBatch] with Arm {
+    spillCallback: RapidsBuffer.SpillCallback)
+  extends Iterator[(ColumnarBatch, ColumnarBatch)] with Arm {
 
-  // Currently do it in a somewhat ugly way. In the future cuDF will provide a dedicated API.
-  // Current solution assumes one group data exists in only one batch, so just split the
-  // batch into multiple batches to make sure one batch contains only one group.
-  private val groupBatches: mutable.Queue[SpillableColumnarBatch] = mutable.Queue.empty
+  private val queueBatches: mutable.Queue[(SpillableColumnarBatch, Seq[Int])] =
+    mutable.Queue.empty
 
   TaskContext.get().addTaskCompletionListener[Unit]{ _ =>
-    groupBatches.foreach(_.close())
-    groupBatches.clear()
+    queueBatches.foreach(_._1.close())
+    queueBatches.clear()
   }
 
-  override def hasNext(): Boolean = groupBatches.nonEmpty || wrapped.hasNext
+  override def hasNext(): Boolean = queueBatches.nonEmpty || wrapped.hasNext
 
-  override def next(): ColumnarBatch = {
-    if (groupBatches.nonEmpty) {
-      groupBatches.dequeue().getColumnarBatch()
+  // Return both the original batch and the additional info batch.
+  // The info batch will be used by the Python side.
+  override def next(): (ColumnarBatch, ColumnarBatch) = {
+    val (nextBatch, gpLens) = if (queueBatches.nonEmpty) {
+      val (sBatch, lens) = queueBatches.dequeue()
+      (sBatch.getColumnarBatch(), lens)
     } else {
       val batch = wrapped.next()
       inputBatches += 1
       inputRows += batch.numRows()
-      if (batch.numRows() > 0 && batch.numCols() > 0 && partitionSpec.nonEmpty) {
-        val partitionIndices = partitionSpec.indices
-        // 1) Calculate the split indices via cudf.Table.groupBy and aggregation Count.
-        //   a) Compute the count number for each group in a batch, including null values.
-        //   b) Restore the original order (Ascending with NullsFirst) defined in the plan,
-        //      since cudf.Table.groupBy will break the original order.
-        //   c) The 'count' column can be used to calculate the indices for split.
-        val cntTable = withResource(GpuProjectExec.project(batch, partitionSpec)) { projected =>
-          withResource(GpuColumnVector.from(projected)) { table =>
-            table
-              .groupBy(partitionIndices:_*)
-              .aggregate(Aggregation.count(true).onColumn(0))
-          }
+      val groupLens = genGroupInfoFromBatch(batch)
+      if (isSplitGroups && groupLens.size > 1) {
+        // Split the groups when being required and there are more than one group.
+        val groupBatchesAndSizes = splitBatchPerGroups(batch, groupLens)
+          .zip(groupLens.map(Seq(_)))
+        val tails = groupBatchesAndSizes.tail.map {
```

Comment:
Again a comment here would be helpful. Having the tail on the groupBatchesAndSizes fells wrong because it looks like we are dropping some data, and it is not clear that we have to go down the head to be returned.  It also looks like there might be some leaks on failure, but that is minor compared to having the code be readable.